fairlearn/metrics/_metric_frame.py

```diff
@@ -0,0 +1,482 @@
+        if cf_list is not None:
+            # Prepend the conditional features, so they are 'higher'
+            rows = copy.deepcopy(cf_list) + rows
+
+        return self._compute_dataframe_from_rows(func_dict, y_true, y_pred, rows)
+
+    def _compute_dataframe_from_rows(self, func_dict, y_true, y_pred, rows):
+        if len(rows) == 1:
+            row_index = pd.Index(data=rows[0].classes, name=rows[0].name)
+        else:
+            row_index = pd.MultiIndex.from_product([x.classes for x in rows],
+                                                   names=[x.name for x in rows])
+
+        if len(row_index) > _SUBGROUP_COUNT_WARNING_THRESHOLD:
+            msg = _SUBGROUP_COUNT_WARNING.format(len(row_index))
+            logger.warning(msg)
+
+        result = pd.DataFrame(index=row_index, columns=func_dict.keys())
+        for func_name in func_dict:
+            for row_curr in row_index:
+                mask = None
+                if len(rows) > 1:
+                    mask = self._mask_from_tuple(row_curr, rows)
+                else:
+                    # Have to force row_curr to be an unary tuple
+                    mask = self._mask_from_tuple((row_curr,), rows)
+
+                # Only call the metric function if the mask is non-empty
+                if sum(mask) > 0:
+                    curr_metric = func_dict[func_name].evaluate(y_true, y_pred, mask)
+                    result[func_name][row_curr] = curr_metric
+        return result
+
+    @property
+    def overall(self) -> Union[pd.Series, pd.DataFrame]:
+        """Return the underlying metrics evaluated on the whole dataset.
+
+        If no ``control_features`` were specified, then this is
+        a :class:`pandas.Series` with one row for each underlying metric.
+
+        If ``control_features`` were specified, then this is a
+        :class:`pandas.DataFrame`, with columns corresponding to the
+        underling metric(s) and the rows indexed by the subgroups of
+        the control features.
+        """
+        return self._overall
+
+    @property
+    def by_group(self) -> pd.DataFrame:
+        """Return the collection of metrics evaluated for each subgroup.
+
+        This :class:`pandas.DataFrame` contains the result of evaluating
+        each underlying metric for each combination of classes in the
+        sensitive and control features. The columns identify
+        the underlying metric, while the rows are indexed by the unqiue
+        combinations of classes found in the sensitive and control
+        features. If control features are present, they are in the
+        'outer' layer of the row indexing.
+
+        If a particular combination of classes was not present in the dataset
+        (likely to occur as more sensitive and control features
+        are specified), then the corresponding entry in the DataFrame will
+        be NaN.
+        """
+        return self._by_group
+
+    @property
+    def control_features(self) -> List[str]:
```

Comment:
this should be either `control_feature_names` or `control_levels`. i much prefer latter for conciseness and also for the context where it is used.

Reply:
I prefer `control_levels`; I started off calling it `control_feature_names` but was asked to change.